---
title: "Bioconductor_Q&A_test"
author: "Simone Avesani, Manuel Tognon"
date: "12/9/2020"
output: html_document
---

# EX1

Given the following IRange ir:
```{r}
##          start       end     width
##      <integer> <integer> <integer>
##   A1         1         3         3
##   A2         3         5         3
##   A3         5         7         3
```
how can you subset ir in order to take the second element?
a) ir[[2]]
b) ir[2]
c) ir["A2"]
d) ir[[2]] or ir["A2"]
e) ir[2] or ir["A2"] X
```{r}
```

# EX2

What is a normal IRanges?

a) is an IRanges with values normalized between 0 and 1
b) is a minimal representation of the IRanges viewed as a set X
c) is a minimal representation of the IRanges viewed as an array
d) is a set with as much ranges from the original IRanges object as possible
e) is an array with as much ranges from the original IRanges object as possible
```{r}
```

# EX3 

What is an Hits object obtained after finding the overlaps between two IRanges?

a) it is a DataFrame which describes the relationship between two IRanges
b) it is a vector containing the overlapping positions
c) it is an array containing the overlapping positions
d) it is a matrix which describes the relationship between two IRanges
e) it is a matrix which describes the position in the genome overlapping an IRanges
```{r}
```

# EX4

Which is the difference between IRanges and GRanges objects?

a) There is no difference
b) GRanges are a wrapper class of IRanges
c) GRanges are like IRanges with the addition of 'strand' information
d) GRanges are like IRanges with the addition of 'strand' and 'chromosome' information
e) Only GRanges can be subset using negative indexing
```{r}
```

# EX5

Given an IRanges ir. What does it happen running 
```{r}
flank(ir, width = 3, start = F, both = T)
```
?

a) It takes the flanking position of width 3 downstream and upstream with respect to the end of the ranges of 
   ir
b) It takes the flanking position of width 3 downstream and upstream with respect to the start of the ranges 
   of ir
c) It takes the flanking position of width 3 downstream with respect to the start of the ranges of ir
d) It takes the flanking position of width 3 upstream with respect to the start of the ranges of ir
e) Resize the ranges in ir to a width of 3
```{r}
```

# EX6 

Which is the R data-structure relying under the Rle class?

a) vector
b) matrix
c) array
d) factor
e) BioStrings
```{r}
```

# EX7

Which kind of strings can be represented using a BioStrings object?

a) only DNA
b) RNA and DNA
c) DNA and aminocidic sequences
d) RNA and aminoacidics sequences
e) DNA, RNA and aminoacidics sequences
```{r}
```

# EX8

What is a consensus matrix?

a) a DataFrame containing the translated DNA codons
b) an array containing the letter probabilty at each position
c) a matrix containing  the letter probability at each position
d) a DataFrame containing the letter probability at each position
e) a matrix containing the translated DNA codons
```{r}
```

# EX9

Which is the resulting object calling coverage() on a GRanges?

a) a named Rle List object with one coverage vector per seqlevel of the GRanges
b) the sequencing depth per ranges in the GRanges
c) a coverage vector between all the sequences
d) a named List object with one coverage vector per seqlvel of the Granges
e) coverage() works on Biostrings and not on GRanges
```{r}
```

# EX10

What is an OrgDb object?

a) is one member of a family of annotation objects
b) is one member of a family of OrgDb objects
c) is data.frame of annotation objects
d) contains only data related to the homo sapiens specie
e) an object to access the database of Bioconductor organization
```{r}
```

# EX11

Create two IRanges objects ir1 and ir2 with random start and end positions (in the range 1:100 and make 
sure that start < end). Then, resize ir1 to a width of 10. Find the overlaps between ir1 and ir2. (Hint: use
sample())
```{r}
```

# EX12

Given the following lines of code, comment what it is happening step by step.
```{r}
plotRanges <- function(x, xlim=x, main=deparse(substitute(x)),
                       col="black",sep=0.5,...){
  height <-1
  if(is(xlim, "Ranges"))
    xlim<-c(min(start(xlim)),max(end(xlim)))
   bins<- disjointBins(IRanges(start(x), end(x)+1))
  plot.new()
  plot.window(xlim, c(0, max(bins)*(height+sep)))
  ybottom <- bins*(sep+height) -height
  rect(start(x)-0.5, ybottom, end(x)+0.5, ybottom+height, col=col,...)
  title(main)
  axis(1)
}
par(mfrow = c(2,1))


ir <- IRanges(start= c(1,3,7,9), end=c(4,4,8,10))
plotRanges(ir)
ir1 <- reduce(ir)
plotRanges(ir1) 
ir2 <- disjoin(ir)
plotRanges(ir2)
```

# EX13

Create a GRanges object gr with random start and end positions (in range 1:100), chromosome and strand 
information.  Then, count how many ranges fall in the forward strand and how many ranges have a width less 
than 10. (Hint: use sample())
```{r}
```

# EX14

Given the GRanges gr1 and gr2, sort gr1 by seqlevel and select only certain ranges from gr1 which overlap 
with gr2.
```{r}
gr1 <- GRanges(seqnames = c("chr11", "chr5", "chr3"), IRanges(c(4, 10, 25), c(18, 15, 32)), 
                           strand=c("-", "+", "+"))
gr2 <- GRanges(seqnames = c("chr11", "chr5", "chr3"), IRanges(c(7, 16, 29), c(21, 31, 31)), 
                           strand=c("-", "+", "+"))

```

# EX15

Given a DNA string s in IUPAC CODE MAP, transform it in the simple DNA alphabet (A, C, G, T). Then, compute 
the reverse complement and translate it into the corresponding aminoacidic sequence. (Hint: use 
IUPAC_CODE_MAP)
```{r}
s <- "NTRKWSD"
```

# EX16

Retrieve all annotation data for "Homo sapiens". Which is the data provider? Which is the rdataclass of the 
resulting object? Which is the taxonomy ID? Which are the source type?
```{r}
```

