---
title: "Multiassay_Q&A_test"
output: html_document
---


## ES1

What is a Summarized Experiment

a) is an R class to store matrices of sperimental results (RNAseq, proteomics) with addiotional metadata (x)
b) is an R class to represent a matrix storing multiple data frames
c) is a Python3.9 class to store matrices of sperimental results (RNAseq, proteomics) with addiotional metadata
d) is an R class where the rows represents the genomic ranges of interests
e) is an R class representing an matrices array of sperimental results


## ES2

Which is the difference between a Summarized Experiment and a Ranged Summarized Experiment

a) SE is an object which can contain a limited number of experimental data while RSE can only contain the user defined range of experimental results
b) There is not difference
c) SE is a child class of RSE
d) RSE contains the Genomic ranges in the rows along with the experimental data while the SE object does not contain Genomic ranges objects
e) SE object can contain metadata while RSE cannot contain metadata

## ES3 

What is a Multiassay object:

a) is an R data structure for representing and analyzing transcriptomics and proteomics experiments
b) is an R data structure for representing and analyzing multi-omics experiments
c) is an R data structure for representing metadata of a SE
d) is an R object reprenting an array of SE
e) it is a parent class of SE 

## ES4

What is difference between ColData and RowData in a Multiassay

a) ColData stores specimen metadata while RowData contains information related to experimental data 
b) ColData contains information related to experimental data  while RowData sotores specimen metadata
c) RowData contains the matrices with experimental data while ColData stores the sample identifiers
d) There is not difference
e) ColData is a vector while RowData is a data frame

## ES5

What is a TCGA barcode:

a) identifies cancer gene on NCBI 
b) is a primary identifier of biospecimen data within the TCGA project and it is composed by a single and unique identifier
c) is a primary identifier of biospecimen data within the TCGA project and it is composed by a collection of identifiers (x)
d) identifies the cell line used to obtain TCGA data available on TCGA
e) identifies the protein produced translating the transcript produced by the corresponding gene


## ES6

Import the MultiAssayExperiment miniACC, keep only patients that have complete data for all assay.
Print a table to show how many patiens have complete data for all assay.
Select only the RNASeq2GeneNorm assay and create a completed SummarizedExperiments (adding also the correct colData) using as colnames the TCGAbarcode of the partecipants.




```{r}
library(MultiAssayExperiment)
library(TCGAutils)

sampleMap(miniACC)

data(miniACC)
summary(complete.cases(miniACC))
acc <- miniACC[, complete.cases(miniACC), ]

rna= experiments(acc)[[1]] 

  
colnames(rna)=TCGAbarcode(colnames(rna), participant = TRUE)
rna = rna[,!duplicated(colnames(rna))]


rownames(coldata)
colnames(assay(rna))
se = SummarizedExperiment(assay = list(count=assay(rna)), 
                          colData = colData)

```
## ES7

Consider "airway" dataset, how many patients are treated and how many are untreated?
Which are the uniques names of the cell lines used in the study? (Hint: import airway dataset using library())
Which is the publication from which "airway" data have been obtained?
## ES8

Consider "miniACC" dataset, how many RNA counts are available for genes "MAPK14" and "IGFBP2"?
Retrive the name of the protein produced by "IGFBP2" (Hint: the information is available in miniACC multiassay)


## ES9

Consider the file "pathaways_scoring.tsv", it contains a matrix with the enrichment score of 300 biological pathways into the "airway_new" dataset samples.
Renames the sample names using the identifiers used in "airway_new" dataset, create a summarized experiment and add to it all the  information about of the samples available in the right location.
Compute the average enrichment score by sample and than by pathway. Which are the pathways and samples with highest enrichment score? Put this information in metadata.

Consider the dataset "airway_new.RDS", load it (Hint: use readRDS()),  map the ENSEMBL gene names to the corresponding gene symbols and put the information in the right location of the SummarizedExperiment. 
Retrieve the gene(s) in the treated samples with the highest and lowest average RNA counts and do the same for genes of untreated samples. Put this information in metadata.

Consider the two preovius SummarizedExperiments obtained and create a MultiAssay containing them.
```{r}
library("airway")
airway_new <- airway[1:3000,]
rownames(airway_new)
BiocManager::install("EnsDb.Hsapiens.v79")
library(EnsDb.Hsapiens.v79)
geneIDs1 <- ensembldb::select(EnsDb.Hsapiens.v79, keys= rownames(airway_new), keytype = "GENEID", columns = c("SYMBOL","GENEID"))
any(is.na(geneIDs1$SYMBOL))

```

